File: server/routes/farm.js
const express = require('express');
const router = express.Router();
const db = require('../database');
const { generateRecommendation } = require('../engine/recommender');

// GET /farm/:id/status
router.get('/:id/status', (req, res) => {
    const farmId = req.params.id;

    // We need to fetch: Farm Details, Latest IoT, Latest Drone, KnowledgeBase
    const queries = {
        farm: "SELECT * FROM farms WHERE farm_id = ?",
        iot: "SELECT * FROM iot_readings WHERE farm_id = ? ORDER BY timestamp DESC LIMIT 1",
        drone: "SELECT * FROM drone_analysis WHERE farm_id = ? ORDER BY analysis_date DESC LIMIT 1",
        kbCrops: "SELECT * FROM kb_crops",
        kbDiseases: "SELECT * FROM kb_diseases"
    };

    db.get(queries.farm, [farmId], (err, farm) => {
        if (err || !farm) return res.status(404).json({ error: 'Farm not found' });

        db.get(queries.iot, [farmId], (err, iot) => {
            db.get(queries.drone, [farmId], (err, drone) => {
                db.all(queries.kbCrops, [], (err, kbCrops) => {
                    db.all(queries.kbDiseases, [], (err, kbDiseases) => {

                        // Generate Recommendation
                        const rec = generateRecommendation(farm, iot, drone, kbCrops, kbDiseases);

                        res.json({
                            farm,
                            latest_iot: iot || null,
                            latest_drone: drone || null,
                            recommendation: rec
                        });

                    });
                });
            });
        });
    });
});

// GET /farm/user/:userId/status
router.get('/user/:userId/status', (req, res) => {
    const userId = req.params.userId;

    // Fetch farm associated with user
    db.get("SELECT * FROM farms WHERE user_id = ?", [userId], (err, farm) => {
        if (err || !farm) {
            // If no farm found (shouldn't happen for new users), return empty/defaults
            return res.status(404).json({ error: 'Farm not found for this user' });
        }

        const farmId = farm.farm_id;
        const queries = {
            iot: "SELECT * FROM iot_readings WHERE farm_id = ? ORDER BY timestamp DESC LIMIT 1",
            drone: "SELECT * FROM drone_analysis WHERE farm_id = ? ORDER BY analysis_date DESC LIMIT 1",
            kbCrops: "SELECT * FROM kb_crops",
            kbDiseases: "SELECT * FROM kb_diseases"
        };

        db.get(queries.iot, [farmId], (err, iot) => {
            db.get(queries.drone, [farmId], (err, drone) => {
                db.all(queries.kbCrops, [], (err, kbCrops) => {
                    db.all(queries.kbDiseases, [], (err, kbDiseases) => {

                        // Generate Recommendation
                        const rec = generateRecommendation(farm, iot, drone, kbCrops, kbDiseases);

                        res.json({
                            farm,
                            latest_iot: iot || null,
                            latest_drone: drone || null,
                            recommendation: rec
                        });

                    });
                });
            });
        });
    });
});

module.exports = router;

File: server/routes/farm.py
import logging
from fastapi import APIRouter, Response

from database import db_get, db_all
from engine.recommender import generate_recommendation

router = APIRouter()
logger = logging.getLogger("agro-backend")

@router.get("/{farm_id}/status")
def get_farm_status(farm_id: int, response: Response):
    try:
        farm = db_get("SELECT * FROM farms WHERE farm_id = ?", [farm_id])
        if not farm:
            response.status_code = 404
            return {"error": "Farm not found"}
            
        iot = db_get("SELECT * FROM iot_readings WHERE farm_id = ? ORDER BY timestamp DESC LIMIT 1", [farm_id])
        drone = db_get("SELECT * FROM drone_analysis WHERE farm_id = ? ORDER BY analysis_date DESC LIMIT 1", [farm_id])
        kb_crops = db_all("SELECT * FROM kb_crops")
        kb_diseases = db_all("SELECT * FROM kb_diseases")
        
        rec = generate_recommendation(farm, iot, drone, kb_crops, kb_diseases)
        
        return {
            "farm": farm,
            "latest_iot": iot or None,
            "latest_drone": drone or None,
            "recommendation": rec
        }
    except Exception as e:
        logger.error(f"Error getting farm status: {e}")
        response.status_code = 500
        return {"error": "Internal server error"}

@router.get("/user/{user_id}/status")
def get_user_status(user_id: int, response: Response):
    try:
        farm = db_get("SELECT * FROM farms WHERE user_id = ?", [user_id])
        if not farm:
            response.status_code = 404
            return {"error": "Farm not found for this user"}
            
        farm_id = farm.get("farm_id")
        
        iot = db_get("SELECT * FROM iot_readings WHERE farm_id = ? ORDER BY timestamp DESC LIMIT 1", [farm_id])
        drone = db_get("SELECT * FROM drone_analysis WHERE farm_id = ? ORDER BY analysis_date DESC LIMIT 1", [farm_id])
        kb_crops = db_all("SELECT * FROM kb_crops")
        kb_diseases = db_all("SELECT * FROM kb_diseases")
        
        rec = generate_recommendation(farm, iot, drone, kb_crops, kb_diseases)
        
        return {
            "farm": farm,
            "latest_iot": iot or None,
            "latest_drone": drone or None,
            "recommendation": rec
        }
    except Exception as e:
        logger.error(f"Error getting user farm status: {e}")
        response.status_code = 500
        return {"error": "Internal server error"}
